### DISCUSSION TITLE
.gitignore and /vendor

### DISCUSSION THREAD

**Discussion by joshlewis** [spatie/wordpress-ray#62] [General]
Hey all! This is my first time posting here, and it makes me a little nervous. Let me know if I'm saying anything foolish.

I encountered a problem yesterday with a recent update to the WordPress plugin for Ray, and I'm not really sure what to think about it. I'll explain what happened, but I'll try not to assign any blame, since I don't really know what "best practices" are in my situation. It's possible I'm out of line.

I use git to update the WordPress sites I work on. The site in this story is hosted on WP Engine, which encourages devs to use git to push their changes to their live sites. 

So, I updated Ray on my local dev system. It was on 1.5.4, and I updated it to 1.7.2. Then I used `git add .` to add all the updated files for Ray to my repo, and `git push production` to push those changes to my production remote. WP Engine's scripts then went about deploying the new version of the site.

The site immediately went down. After digging around, I figured out that it had something to do with the fact that [`vendor` is excluded in the `.gitignore` file](https://github.com/spatie/wordpress-ray/blob/master/.gitignore#L11) at the root of the Ray plugin for WordPress. Because it was excluded, `git add .` didn't see the changes to that directory and didn't add it, and when I pushed the changes to production, I was shipping incomplete code, and WordPress went down entirely.

I noticed that `.distignore` says `# DO NOT EXCLUDE /vendor` on line 16, so clearly I should have added the vendor directory into my git repo in my case. But git doesn't look at `.distignore` (or at least my version doesn't?), and I didn't notice that line until it was too late.

There are lots of things I could have done differently to prevent the site going down, so I'm clearly not without blame here! 

Here's my question: is excluding _necessary_ files using the `.gitignore` file compatible with a git-based shipping process without requiring the dev to manually look for `.distignore` files and read all their changes every time the plugin changes?

The problem wouldn't have reproduced on my local system since WordPress installed the plugin and it included the vendor directory. I could have noticed the issue if I had deployed to a staging environment first. Other than that, am I missing some "best practices" steps I should be taking to prevent this kind of thing from happening? Is there some change I should make to my basic `git add .` process when updating plugins?

I generally don't attempt to run `composer install` on every plugin every time it updates. I guess I'm hoping that doesn't need to become part of my process, since it's not really a regular part of how WordPress works.

Thank you so much for your consideration! I appreciate any advice you choose to give.

**Comment by joshlewis** [comment]
One other thought I had: I assume if the `.distignore` file had contained a line for `.gitignore`, this problem would not have happened, since in that situation, `.gitignore` would not have been present in the downloaded plugin, and the standard `git add .` would have picked up the vendor directory in that case.

Would it make sense to add `.gitignore` to `.distignore`?

**Comment by jeffreyvr** [comment]
Hi @joshlewis, first of all, there is nothing foolish about your post!

I think your suggestion about adding `.gitignore` to `.distignore` makes sense. I think that would indeed solve the problem, and the `.gitignore` file should not be necessary in the distributed version of the plugin. 

I created a PR that implements this: https://github.com/spatie/wordpress-ray/pull/63

**Reply by freekmurze** [reply]
Thank you both!
